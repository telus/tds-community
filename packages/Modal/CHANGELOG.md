# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.6.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.5.0...@tds/community-modal@1.6.0) (2021-09-21)


### Features

* **community-modal:** react 17 and styled components 5 ([17029bb](https://github.com/telus/tds-community/commit/17029bba5ab26265653fc7a00d6f2ff46135f4e7))





# [1.5.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.4.0...@tds/community-modal@1.5.0) (2021-07-21)


### Features

* **community-modal:** adjusted height and margins around modal ([86bcb1e](https://github.com/telus/tds-community/commit/86bcb1e40f16532df657cf511d1679d82862063a))
* **community-modal:** applied changes requested in code review ([8fd0cdc](https://github.com/telus/tds-community/commit/8fd0cdc7ecd8b4f6cc7b203bea979b8be557e93c))
* **community-modal:** applied max-height to modal overlay ([c5b27d1](https://github.com/telus/tds-community/commit/c5b27d13382f793371158e2c3fc1176dabc753bf))
* **community-modal:** fix unit tests ([b1309a0](https://github.com/telus/tds-community/commit/b1309a0cd5b1662e55f27c7cb642c5b2d7c50b5e))
* **community-modal:** fixed content-wrapper styling issue ([111f7fb](https://github.com/telus/tds-community/commit/111f7fb8ed5f760b62a6441a93ba1d74dcef12ba))
* **community-modal:** fixed styling issues in mobile devices ([69a09d0](https://github.com/telus/tds-community/commit/69a09d0912078e9f14207527246ac0f3571364bc))
* **community-modal:** fixed unit test snapshot ([6ebda37](https://github.com/telus/tds-community/commit/6ebda37125ed4bab902dbb0d547a09b972b354db))
* **community-modal:** removed notification from header ([1cd66e3](https://github.com/telus/tds-community/commit/1cd66e3cc634536ad96bfcbbe6bf248bf51778eb))
* **community-modal:** updated dependency files ([90bcb94](https://github.com/telus/tds-community/commit/90bcb94a6b3b82d45e3d1b19501b68369436c52a))
* **community-modal:** updated documentation for props ([bbd8e57](https://github.com/telus/tds-community/commit/bbd8e57a26eaf81b7316ef77f7a743c60858bc5e))
* **community-modal:** updated tabindex for required components ([c8cbd06](https://github.com/telus/tds-community/commit/c8cbd06b3c0c4f57ee84aaf3fa4d06e6b210fec4))
* **community-modal:** updated unit test snapshots ([94900c3](https://github.com/telus/tds-community/commit/94900c343a34b0c55cbf77c68c631dc7df42d5e1))
* **community-modal:** upgrade modal to match new wireframes ([47350b1](https://github.com/telus/tds-community/commit/47350b13ef277238898524219e536927957b0e0b))





# [1.4.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.3.0...@tds/community-modal@1.4.0) (2021-06-30)


### Features

* **Modal:** pass event to onClose ([8f74286](https://github.com/telus/tds-community/commit/8f742865352a67e47c1feb350f9a1d3e0e89418c))





# [1.3.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.2.2...@tds/community-modal@1.3.0) (2021-03-18)


### Bug Fixes

* **community-modal:** update docs ([7d36e81](https://github.com/telus/tds-community/commit/7d36e81d971d33c4180d24d89dab5eced1f16939))


### Features

* **community-modal:** update modal max-width ([4186943](https://github.com/telus/tds-community/commit/41869438f4026e515f6b1bc58ca99644d894e5e9))





## [1.2.2](https://github.com/telus/tds-community/compare/@tds/community-modal@1.2.1...@tds/community-modal@1.2.2) (2021-01-15)


### Bug Fixes

* **community-modal:** [#403](https://github.com/telus/tds-community/issues/403) fixed heading text overlap with close icon ([0765d0d](https://github.com/telus/tds-community/commit/0765d0d387ee7c7ae7c32ad9b807103d528e02bb))





## [1.2.1](https://github.com/telus/tds-community/compare/@tds/community-modal@1.2.0...@tds/community-modal@1.2.1) (2021-01-15)


### Bug Fixes

* **community-modal:** 432 fix for modal bottom cta wrong padding top ([2466a03](https://github.com/telus/tds-community/commit/2466a03c50e0c5bc9a02f2aed8444ee85a4d8836))





# [1.2.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.1.1...@tds/community-modal@1.2.0) (2020-09-30)


### Features

* **community-modal:** make confirmCTAText and onConfirm props optional ([7b45040](https://github.com/telus/tds-community/commit/7b45040ece3d70e8fbeff55c3517e2a020a419ee))





## [1.1.1](https://github.com/telus/tds-community/compare/@tds/community-modal@1.1.0...@tds/community-modal@1.1.1) (2020-09-09)


### Bug Fixes

* **community-modal:** update util-helpers version ([78e9d46](https://github.com/telus/tds-community/commit/78e9d46d049b669d03c763524dc0e959b7f3c79d))





# [1.1.0](https://github.com/telus/tds-community/compare/@tds/community-modal@1.0.0...@tds/community-modal@1.1.0) (2020-07-10)


### Bug Fixes

* **community-modal:** handling custom component with same parameters ([cf8fe75](https://github.com/telus/tds-community/commit/cf8fe75eaa496c9bb488a7e5680d6442f2925e51))
* **community-modal:** handling custom component with same parameters ([1d4d0bd](https://github.com/telus/tds-community/commit/1d4d0bd3096724e634040e9c15cf56ae95c2709c))
* **community-modal:** revert package number ([0098d2f](https://github.com/telus/tds-community/commit/0098d2f8948554b658e369e3536408bef9db1362))


### Features

* **community-modal:** functionality to customize description ([587acd0](https://github.com/telus/tds-community/commit/587acd0088c502bc76145e4ccd0670c57d8131d1))





# 1.0.0 (2020-04-13)


### Features

* **community-modal:** new component ([39ba653](https://github.com/telus/tds-community/commit/39ba65325b701757f07554404983c33d227a2fe8))
