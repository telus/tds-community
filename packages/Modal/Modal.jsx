import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'

import Box from '@tds/core-box'
import Button from '@tds/core-button'
import Paragraph from '@tds/core-paragraph'
import { safeRest } from '@tds/util-helpers'
import { Close, IconButton } from '@tds/core-interactive-icon'
import Heading from '@tds/core-heading'
import { withFocusTrap } from '@tds/shared-hocs'

import {
  StyledModal,
  CTAWrapper,
  CloseButtonWrapper,
  FullScreenOverlay,
  ModalWrapper,
  PaddingOverride,
  OutlineButton,
} from './styles'

const FocusTrap = withFocusTrap('div')

/**
 * @version ./package.json
 */
const Modal = ({
  heading,
  confirmCTAText,
  cancelCTAText,
  bodyText,
  isOpen,
  onClose,
  onConfirm,
  focusElementAfterClose,
  ...rest
}) => {
  const ModalOverlayRef = useRef(null)
  const modalRef = useRef(null)
  const header = useRef(null)
  const firstCTA = useRef(null)
  const closeButton = useRef(null)

  const handleClose = () => {
    if (focusElementAfterClose && focusElementAfterClose.current) {
      focusElementAfterClose.current.focus()
    }
    return onClose()
  }

  const handleKeyDown = e => {
    const key = e.keyCode || e.key
    if (key === 'Escape' || key === 27) {
      return handleClose()
    }
    return null
  }

  const preventScroll = e => {
    if (ModalOverlayRef.current.contains(e.targetTouches[0].target)) {
      e.preventDefault()
    }
  }

  const removeEventScrolling = () => {
    document.body.removeEventListener('touchmove', preventScroll)
    document.body.style.overflow = 'visible'
  }

  const handleOutSideClick = e => {
    if (!modalRef.current.contains(e.target)) {
      handleClose()
    }
  }

  useEffect(() => {
    if (isOpen) {
      document.body.addEventListener('touchmove', preventScroll, { passive: false })
      document.body.addEventListener('keydown', handleKeyDown, { passive: false })
      document.body.addEventListener('mousedown', handleOutSideClick, { passive: false })
      document.body.style.overflow = 'hidden'
      header.current.focus()
    } else {
      removeEventScrolling()
      document.body.removeEventListener('keydown', handleKeyDown)
      document.body.removeEventListener('mousedown', handleOutSideClick)
    }
    return () => {
      removeEventScrolling()
      document.body.removeEventListener('keydown', handleKeyDown)
      document.body.removeEventListener('mousedown', handleOutSideClick)
    }
  }, [isOpen])

  return (
    <React.Fragment>
      {isOpen && (
        <FocusTrap>
          <FullScreenOverlay
            data-testid="tds-modal-overlay"
            {...safeRest(rest)}
            isOpen={isOpen}
            ref={ModalOverlayRef}
          >
            <StyledModal ref={modalRef}>
              <CloseButtonWrapper>
                <IconButton icon={Close} onClick={handleClose} ref={closeButton} a11yText="Close" />
              </CloseButtonWrapper>

              <ModalWrapper>
                <Box inset={5}>
                  <Box between={3}>
                    <div ref={header} tabIndex="-1">
                      <Heading level="h3" tag="div">
                        {heading}
                      </Heading>
                    </div>
                    <Paragraph>{bodyText}</Paragraph>
                  </Box>
                  <PaddingOverride>
                    <Box vertical={5}>
                      <CTAWrapper cancelCTAExists={cancelCTAText}>
                        <Button ref={firstCTA} onClick={onConfirm}>
                          {confirmCTAText}
                        </Button>
                        {cancelCTAText && (
                          <OutlineButton ref={closeButton} onClick={handleClose}>
                            <Button>{cancelCTAText}</Button>
                          </OutlineButton>
                        )}
                      </CTAWrapper>
                    </Box>
                  </PaddingOverride>
                </Box>
              </ModalWrapper>
            </StyledModal>
          </FullScreenOverlay>
        </FocusTrap>
      )}
    </React.Fragment>
  )
}

Modal.propTypes = {
  /**
   * Text that will appear at the top of the content section.
   */
  heading: PropTypes.string.isRequired,
  /**
   *
   * Text that will appear in the middle of the content section.
   */
  bodyText: PropTypes.string.isRequired,
  /**
   *
   * Text that represents confirm CTA.
   */
  confirmCTAText: PropTypes.string.isRequired,
  /**
   *
   * Text that represents cancel CTA or closing modal action.
   */
  cancelCTAText: PropTypes.string,
  /**
   *
   * Boolean variable that controls open and closed state of modal.
   */
  isOpen: PropTypes.bool.isRequired,
  /**
   *
   * Use this prop to change the `open` when the customer clicks the cancel button, close button, background, or hits the escape key. Do not use this handler to perform account-related actions for the customer.
   */
  onClose: PropTypes.func.isRequired,
  /**
   *
   * Handler Function used to proceed with modal CTA.
   */
  onConfirm: PropTypes.func.isRequired,
  /**
   *
   * Accepts a React Element's Ref, in order to focus it after modal closes.
   */
  focusElementAfterClose: PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.any })])
    .isRequired,
}

Modal.defaultProps = {
  cancelCTAText: '',
}

export default Modal
