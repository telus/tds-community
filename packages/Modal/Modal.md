Before you use modals:

- Does the interaction need to be disruptive?
- Have you tried other less disruptive options but they didn’t meet your needs?
- Is the information critical to be consumed before completing the task on this page?
- Are you not able to include the critical question or information within this page?

If you answered yes to the questions above, you may need to use a modal.

A modal window is a secondary window that opens on top of the main one. Users have to interact with it before they can carry out their task and return to the main window. Use to reveal additional information to a user after they have performed an explicit interaction. They are a strongly discouraged pattern; it's preferred to have all relevant information within a page, and irrelevant information either linked externally or omitted.

There are two kinds of modals, [Content Modal](#tds-content-modal) and [Dialog Modal](#tds-dialog-modal).

### Usage criteria

- Must only appear after a customer interaction, not on page load or any other circumstance
- Open a modal based on explicit customer action e.g. clicking on a button/link/form field
- Must only contain copy and no other content types
- Should keep copy to a minimum
- Only one modal should be "current" at any time
- Refer to [WebAIM documentation](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html) for how to achieve modals accessibly
- Don’t use modals to reinforce or repeat information already available in the parent page or view
- Don’t use modals consecutively

### Accessibility features

- When displayed, the modal focus-traps the user, and locks page scroll
- It is scrollable when the viewport height is shorter than the contents of the modal

<div id="tds-content-modal"></div>
### Content Modal

A content modal provides a way for the customer to access additional information without leaving the current page or view, and maintain context. It’s also an effective technique to catch the customer's attention to something vital, a specific task, or set of information.

```jsx
const ExampleContentModal = () => {
  const [showModal, setShowModal] = React.useState(false)
  const focusElementOnClose = React.useRef(null)

  const openModal = () => {
    setShowModal(true)
  }

  const closeModal = () => {
    setShowModal(false)
  }

  const proceed = () => {
    alert('modal proceed')
    setShowModal(false)
  }

  return (
    <div>
      <Modal
        heading="Make a statement for a content modal"
        bodyText="Use a content modal for dismissible content. A content modal provides a way for the user to access additional information without leaving the current page or view, and maintain context"
        confirmCTAText="Primary action"
        focusElementAfterClose={focusElementOnClose}
        onConfirm={proceed}
        onClose={closeModal}
        isOpen={showModal}
      />
      <Box inset={4} between={3}>
        <Heading level="h2">Content Modal</Heading>
        <div>
          <Button ref={focusElementOnClose} onClick={openModal}>
            Open Modal
          </Button>
        </div>
      </Box>
    </div>
  )
}
;<ExampleContentModal />
```

<div id="tds-dialog-modal"></div>
### Dialog Modal

A dialog modal provides information to the customer and prompts them for a response.

### Usage criteria for Dialog Modals

- Use dialog modals as a means for customers to safely make actions that require their full attention, such as confirmations of destructive actions
- The copy on the modal should clearly lead to confirm or reject an impending decision and if possible the outcomes
- Limit content to less than 150 characters for the title, and less than 400 characters for body text
- Avoid dialog modals:
  - if there are more than 2 options for customer consideration
  - as alerts to display errors, warnings, and success messages
  - being shown consecutively

```jsx
const ExampleDialogModal = () => {
  const [showModal, setShowModal] = React.useState(false)
  const focusElementOnClose = React.useRef(null)

  const openModal = () => {
    setShowModal(true)
  }

  const closeModal = () => {
    setShowModal(false)
  }

  const proceed = () => {
    alert('modal proceed')
    setShowModal(false)
  }

  return (
    <div>
      <Modal
        heading="Ask a question that confirms the customer action which initiated the dialog modal"
        bodyText="Use a dialog modal to provide information and prompt for a response to customer action, which may lead to a significant impact on account or lead to unrecoverable states."
        confirmCTAText="Affrimative action"
        cancelCTAText="Negative action"
        focusElementAfterClose={focusElementOnClose}
        onConfirm={proceed}
        onClose={closeModal}
        isOpen={showModal}
      />
      <Box inset={4} between={3}>
        <Heading level="h2">Dialog Modals</Heading>
        <div>
          <Button ref={focusElementOnClose} onClick={openModal}>
            Open Modal
          </Button>
        </div>
      </Box>
    </div>
  )
}
;<ExampleDialogModal />
```
