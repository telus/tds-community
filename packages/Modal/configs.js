export const HEADER_LEVELS = {
  H1: 'h1',
  H2: 'h2',
  H3: 'h3',
  H4: 'h4',
}

export const TEXT_SIZES = {
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large',
}

export const BUTTON_RANKS = {
  MAIN: 'main',
  COMMON: 'common',
}

export const CANCELLATION_BUTTON_TYPES = {
  BUTTON: 'button',
  LINK_WITH_ICON: 'linkWithIcon',
  LINK_WITHOUT_ICON: 'linkWithoutIcon',
}

export const BUTTON_VARIANTS = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  INVERTED: 'inverted',
  STANDARD: 'standard',
  BRAND: 'brand',
  DANGER: 'danger',
}

export const NOTIFICATION_VARIANTS = {
  INSTRUCTIONAL: 'instructional',
  BRANDED: 'branded',
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
}

export const NOTIFICATION_COPY = {
  EN: 'en',
  FR: 'fr',
}

export const LINK_ICON_POSITIONS = {
  LEFT: 'left',
  RIGHT: 'right',
}
