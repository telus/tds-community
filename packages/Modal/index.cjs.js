const Modal = require('./dist/index.cjs')

module.exports = Modal
