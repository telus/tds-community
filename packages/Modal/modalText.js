const copyDictionary = {
  en: {
    confirm: 'Yes',
    cancel: 'No',
  },
  fr: {
    confirm: 'Oui',
    cancel: 'Non',
  },
}
export default copyDictionary
