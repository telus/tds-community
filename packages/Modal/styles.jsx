import styled from 'styled-components'
import { media } from '@tds/core-responsive'
import { colorWhite, colorAccessibleGreen } from '@tds/core-colours'

export const FullScreenOverlay = styled.div(props => {
  if (props && props.modalOpen) {
    return {
      overflow: 'scroll',
      backgroundColor: colorWhite,
      width: '100%',
      height: '100%',
      position: 'fixed',
      top: '0',
      left: '0',
      zIndex: '1400',
      ...media.from('md').css({
        backgroundColor: 'rgba(255, 255, 255, 0.94)',
      }),
    }
  }
  return null
})

export const StyledModal = styled.div({
  height: '100%',
  width: '100%',
  position: 'relative',
  top: '0%',
  left: '0%',
  '> button:first-child': {
    display: 'flex',
    alignSelf: 'flex-end',
    background: 'none',
    border: 'none',
    margin: '1rem 0.5rem',
  },
  ...media.from('md').css({
    margin: '0 auto',
    maxWidth: '570px',
    height: 'auto',
    borderRadius: '0.25rem',
    boxShadow: '0 0 16px 0 rgba(0,0,0,0.2)',
    zIndex: '2000',
    backgroundColor: colorWhite,
    top: '29%',
  }),
})

export const CTAWrapper = styled.div(props => {
  if (props && !props.cancelCTAExists) {
    return {
      ...media.from('md').css({
        display: 'flex',
        flexFlow: 'row',
        '> button:first-child': {
          marginRight: '1rem',
        },
      }),
    }
  }
  return {
    display: 'flex',
    flexFlow: 'column',
    alignItems: 'baseline',
    '> button:first-child': {
      marginBottom: '1rem',
    },
    ...media.from('md').css({
      flexFlow: 'row',
      '> button:first-child': {
        marginRight: '1rem',
        marginBottom: '0rem',
      },
    }),
  }
})

export const PaddingOverride = styled.div({
  '> div:first-child': {
    paddingBottom: '0rem',
  },
})

export const OutlineButton = styled.button({
  background: 'none',
  border: 'none',
  padding: '0rem',
  width: '100%',
  ...media.from('md').css({
    width: 'auto',
    minWidth: '180px',
  }),
  '> button: first-child': {
    color: colorAccessibleGreen,
    background: colorWhite,
    border: `1px solid ${colorAccessibleGreen}`,
    outline: 'none',
    ':hover': {
      background: colorAccessibleGreen,
      color: colorWhite,
    },
  },
})

export const CloseButtonWrapper = styled.div({
  float: 'right',
  padding: '1rem 1rem 0 0',
  position: 'absolute',
  right: '0%',
  top: '0%',
})

export const ModalWrapper = styled.div({
  display: 'flex',
  flexFlow: 'column',
  justifyContent: 'space-between',
  overflow: 'scroll',
  height: '100%',
  '> div:first-child': {
    height: '100%',
    display: 'flex',
    flexFlow: 'column',
    justifyContent: 'space-between',
  },
})
