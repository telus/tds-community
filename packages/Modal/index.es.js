import Modal from './dist/index.es'

export default Modal
