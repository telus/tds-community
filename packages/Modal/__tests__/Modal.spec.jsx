import React from 'react'
import { shallow } from 'enzyme'

import Modal from '../Modal'

describe('Modal', () => {
  const doShallow = (props = {}) => shallow(<Modal {...props} />)

  it('renders', () => {
    const modal = doShallow()

    expect(modal).toMatchSnapshot()
  })

  it('does other things', () => {
    const modal = doShallow()

    expect(modal).toExist()
  })

  it('passes additional attributes to the element', () => {
    const modal = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(modal).toHaveProp('id', 'the-id')
    expect(modal).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const modal = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(modal).not.toHaveProp('className', 'my-custom-class')
    expect(modal).not.toHaveProp('style')
  })
})
