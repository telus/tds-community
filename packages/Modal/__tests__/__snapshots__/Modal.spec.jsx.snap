// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders 1`] = `
<Modal
  bodyText="Are you sure?"
  cancelCTAText=""
  confirmCTAText="I am sure"
  heading="A heading"
  modalOpen={true}
  proceedModalHandler={[Function]}
/>
`;
